kaiax/valset/impl/getter.go

```diff
@@ -0,0 +1,211 @@
+		}
+		return c.selectRandomCommittee(valCtx, round, proposers)
+	}
+	return c.selectRandaoCommittee(valCtx.prevBlockResult)
+}
+
+// GetProposer picks a proposer for the (N, Round) view. it is picked from different sources.
+// - if the policy defaultSet, the proposer is picked from the council. it's defaultSet, so there's no demoted.
+// - if the policy is weightedrandom and before Randao hf, the proposer is picked from the proposers
+// - if the pilicy is weightedrandom and after Randao hf, the proposer is picked from the committee
+func (v *ValsetModule) GetProposer(num uint64, round uint64) (common.Address, error) {
+	// if the block number is genesis, directly return proposer as the first element of councilAddressList.
+	if num == 0 {
+		councilAddressList, err := ReadCouncilAddressListFromDb(v.ChainKv, 0)
+		if err != nil {
+			return common.Address{}, err
+		}
+		return councilAddressList[0], nil
+	}
+
+	valCtx, err := newValSetContext(v, num)
+	if err != nil {
+		return common.Address{}, err
+	}
+	c, err := newCouncil(v.ChainKv, valCtx)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	var (
+		proposerPolicy = valCtx.prevBlockResult.proposerPolicy
+		rules          = valCtx.rules
+	)
+	if proposerPolicy.IsDefaultSet() {
+		lastProposerIdx := c.councilAddressList.getIdxByAddress(valCtx.prevBlockResult.author)
+		seed := defaultSetNextProposerSeed(proposerPolicy, valCtx.prevBlockResult.author, lastProposerIdx, round)
+		idx := int(seed) % len(c.councilAddressList)
+		return c.councilAddressList[idx], nil
+	}
+
+	// before Randao, weightedrandom uses proposers to pick the proposer.
+	// the proposers have always been calculated before entering proposer
+	if !rules.IsRandao {
+		proposers, err := v.getProposers(num)
+		if err != nil {
+			return common.Address{}, err
+		}
+		proposer, proposerIdx := c.getProposerFromProposers(proposers, round)
+		if proposerIdx == -1 {
+			return common.Address{}, errUnknownProposer
+		}
+		return proposer, nil
+	}
+
+	// after Randao, pick proposer from randao committee.
+	committee, err := c.selectRandaoCommittee(valCtx.prevBlockResult)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	idx := int(round) % len(c.qualifiedValidators)
+	proposer, proposerIdx := committee[idx], idx
+
+	// for non-default, fall-back to roundrobin proposer
+	if proposerIdx == -1 {
+		logger.Warn("Failed to select a new proposer, thus fall back to roundRobinProposer")
+		idx = c.qualifiedValidators.getIdxByAddress(proposer)
+		seed := defaultSetNextProposerSeed(params.Sticky, valCtx.prevBlockResult.author, idx, round)
+		proposerIdx = int(seed) % len(c.qualifiedValidators)
+		proposer = c.qualifiedValidators[proposerIdx]
+	}
+	return proposer, nil
+}
+
+func defaultSetNextProposerSeed(policy ProposerPolicy, proposer common.Address, proposerIdx int, round uint64) uint64 {
+	seed := round
+	if proposerIdx > -1 {
+		seed += uint64(proposerIdx)
+	}
+	if policy == params.RoundRobin && !common.EmptyAddress(proposer) {
```

Comment:
When is `proposer` empty?

Reply:
I’m not entirely sure. I can’t think of other use cases, but one use case is that this code was copied from the original source (istanbul.Validators) and I believe the empty address check was meant to handle the genesis block case, where a proposer doesn’t exist. However, the kaiax/valSet code seems doesn't calculate the genesis block's proposer, so this check might no longer be necessary. I'll change the code to reject the empty address.